File: javascript/clock/clock.js
export class Clock {
  constructor(hours = 0, minutes = 0) {
    this.minutes = calculateMinutes(minutes)
    this.hours = calculateHours(hours, minutes)
  }

  toString() {
    const hours = this.hours.toString().padStart(2, '0')
    const minutes = this.minutes.toString().padStart(2, '0')

    return (`${hours}:${minutes}`);
  }

  plus(minutes) {
    minutes = this.minutes + minutes

    return(new Clock(this.hours, minutes))
  }

  minus(minutes) {
    minutes = this.minutes - minutes

    return (new Clock(this.hours, minutes))
  }

  equals(clock) {
    return (this.toString() == clock.toString())
  }
}

const calculateMinutes = (minutes) => {
  // ruby mod does the rolling over for you
  // JS seems more logical and will give a negative answer for a negative argument
  // Hence this wee rolling routine
  while (minutes < 0) {
    minutes += 60
  }

  return (minutes % 60)
}

const calculateHours = (hours, minutes) => {
  // JS version of integer division
  hours += Math.floor(minutes / 60)


  while (hours < 0) {
    hours += 24
  }

  // -1 % 24 = -1
  return (hours % 24)
}

File: ruby/clock/clock.rb
class Clock
  attr_reader :hours, :minutes

  def initialize(hour: 0, minute: 0)
    @minutes = calculate_minutes(minute)
    @hours = calculate_hours(hour, minute)
  end

  def to_s
    "%02d:%02d" % [hours, minutes]
  end

  def + (clock)
    hour = hours + clock.hours
    minute = minutes + clock.minutes

    Clock.new(hour: hour, minute: minute)
  end

  def - (clock)
    hour = hours - clock.hours
    minute = minutes - clock.minutes

    Clock.new(hour: hour, minute: minute)
  end

  def == (clock)
    hours == clock.hours && minutes == clock.minutes
  end

  private

  def calculate_minutes(minute)
    # minutes roll over to zero every 60 minutes
    minute % 60
  end

  def calculate_hours(hour, minute)
    # every 60 minutes, add an hour
    hour += minute / 60

    # hours roll over to zero every 24 hours
    # ruby mod handles the rolling over
    # -1 % 24 = 23
    hour % 24
  end
end